File: en/codes/swift/utils/ListNode.swift
/**
 * File: ListNode.swift
 * Created Time: 2023-01-02
 * Author: nuomi1 (nuomi1@qq.com)
 */

public class ListNode: Hashable {
    public var val: Int // 节点值
    public var next: ListNode? // 后继节点引用

    public init(x: Int) {
        val = x
    }

    public static func == (lhs: ListNode, rhs: ListNode) -> Bool {
        lhs.val == rhs.val && lhs.next.map { ObjectIdentifier($0) } == rhs.next.map { ObjectIdentifier($0) }
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(val)
        hasher.combine(next.map { ObjectIdentifier($0) })
    }

    public static func arrToLinkedList(arr: [Int]) -> ListNode? {
        let dum = ListNode(x: 0)
        var head: ListNode? = dum
        for val in arr {
            head?.next = ListNode(x: val)
            head = head?.next
        }
        return dum.next
    }
}

File: codes/typescript/modules/ListNode.ts
/**
 * File: ListNode.ts
 * Created Time: 2022-12-10
 * Author: Justin (xiefahit@gmail.com)
 */

/* 链表节点 */
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

/* 将数组反序列化为链表 */
function arrToLinkedList(arr: number[]): ListNode | null {
    const dum: ListNode = new ListNode(0);
    let head = dum;
    for (const val of arr) {
        head.next = new ListNode(val);
        head = head.next;
    }
    return dum.next;
}

export { ListNode, arrToLinkedList };
